Source/NSUserDefaults.m

```diff
@@ -1629,9 +1658,20 @@ - (void) setObject: (id)value forKey: (NSString*)defaultName
           [_persDomains setObject: pd forKey: processName];
 	  RELEASE(pd);
 	}
+      // Make sure to search all domains and not only the process domain
+      old = [self objectForKey: defaultName];
       if ([pd setObject: value forKey: defaultName])
         {
           [self _changePersistentDomain: processName];
+          [self _notifyObserversOfChangeForKey: defaultName oldValue:old newValue:value];
```

Comment:
The new value must be fetched from defaults if `value` is nil, as there might e.g. be a registered default.

```suggestion
          id new;
          [self _changePersistentDomain: processName];
          new = value ?: [self objectForKey: defaultName];
          [self _notifyObserversOfChangeForKey: defaultName oldValue:old newValue:new];
```